### PR TITLE
Magic link

### PR DESCRIPTION
**Pull request by jwest60**
[from the mobile repo](https://github.com/statonlab/Treesnap-mobile/issues/549)

added a share URL that will create a token for the given observation, allowing the observation to be shared with accurate location to anyone who has the link

### CODE REVIEW

**Comment by almasaeed2010** [comment]
@remind me to test this tomorrow at 8 am

**Comment by almasaeed2010** [comment]
Doesn’t look like we have the reminder bot setup in this repo

File: resources/assets/js/app.jsx [hunk 1]
```diff
@@ -92,6 +92,7 @@ class App extends Component {
           <Route exact={true} path={'/account/groups'} component={GroupsScene}/>
           <Route exact={true} path={'/account/group/:id'} component={GroupScene}/>
           <Route path={'/observation/:id'} component={ObservationScene}/>
+          <Route path={'/share/observation/:id'} component={ObservationScene}/>
```
Comment (almasaeed2010) [inline resources/assets/js/app.jsx]: why is this needed?

File: tests/Feature/ObservationSharingTest.php [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+        $user = factory(User::class)->create();
+
+        $observation = factory(Observation::class)->create([
+            'user_id' => $user->id,
+        ]);
+
+        $this->actingAs($user);
+
+        $response = $this->get('/web/share/observation/' . $observation->id);
+
```
Comment (almasaeed2010) [inline tests/Feature/ObservationSharingTest.php]: You should also test that the returned url matches your expectation. There are 2 ways to do this. Either read the returned url to extract the token and test that it exists in the share_tokens table or get the token  from the share_tokens table and make sure it exists in the URL. Since the user and observation are created in the test by a factory, it is guaranteed that no other share token exists for the combo user_id and observation_id.

File: tests/Feature/ObservationSharingTest.php [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        $token_value = Str::random(60);
+
+        $token = factory(ShareToken::class)->create([
+            'user_id' => $user->id,
+            'observation_id' => $observation->id,
+            'value' => $token_value,
+        ]);
+
+        $this->actingAs($user);
```
Comment (almasaeed2010) [inline tests/Feature/ObservationSharingTest.php]: Since this is the owner of the observation, this is guaranteed to return the accurate location. You need a different user to own the observation and generate the token. Then, the other user can try to use the token to access accurate location.

Comment (almasaeed2010) [reply tests/Feature/ObservationSharingTest.php]: In fact, even anonymous users can access the observation if they have the token so you don't need to authenticate at all.

File: resources/assets/js/app.jsx (hunk 1 above)
Comment (jwest60) [reply resources/assets/js/app.jsx]: its not, i just removed it

File: tests/Feature/ObservationSharingTest.php (hunk 3 above)
Comment (jwest60) [reply tests/Feature/ObservationSharingTest.php]: updated the code so that the user in this test no longer owns the observation

File: tests/Feature/ObservationSharingTest.php (hunk 2 above)
Comment (jwest60) [reply tests/Feature/ObservationSharingTest.php]: should this be a separate test?

Comment (almasaeed2010) [reply tests/Feature/ObservationSharingTest.php]: I don't think so. It's related to the simple status assertion

Comment (jwest60) [reply tests/Feature/ObservationSharingTest.php]: test has been updated with an assertSee for the token value on the latest branch

File: tests/Feature/ObservationSharingTest.php [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+            'user_id' => $owner->id,
+        ]);
+
+        $token_value = Str::random(60);
+
+        $token = factory(ShareToken::class)->create([
+            'user_id' => $owner->id,
+            'observation_id' => $observation->id,
+            'value' => $token_value,
+            'expired_at' => Carbon::now(),
```
Comment (almasaeed2010) [inline tests/Feature/ObservationSharingTest.php]: Makes sure it's expired:
```suggestion
            'expired_at' => Carbon::now()->subDays(2),
```
